pkg/kube/watcher/crdwatcher/crdwatcher.go

```diff
@@ -0,0 +1,80 @@
+// Copyright Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package crdwatcher
+
+import (
+	"sync"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/tools/cache"
+
+	"istio.io/istio/pkg/config/schema/collections"
+	"istio.io/istio/pkg/kube"
+	"istio.io/pkg/log"
+)
+
+// Controller watches a ConfigMap and calls the given callback when the ConfigMap changes.
+// The ConfigMap is passed to the callback, or nil if it doesn't exist.
+type Controller struct {
+	informer  cache.SharedIndexInformer
+	mutex     sync.RWMutex
+	callbacks []func(name string)
+}
+
+// NewController returns a new CRD watcher controller.
+func NewController(client kube.Client, callbacks ...func(name string)) *Controller {
+	c := &Controller{
+		callbacks: callbacks,
+	}
+
+	crdMetadataInformer := client.MetadataInformer().ForResource(collections.K8SApiextensionsK8SIoV1Customresourcedefinitions.Resource().
+		GroupVersionResource()).Informer()
+	_ = crdMetadataInformer.SetTransform(kube.StripUnusedFields)
+	_, _ = crdMetadataInformer.AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj any) {
+			crd, ok := obj.(*metav1.PartialObjectMetadata)
+			if !ok {
+				// Shouldn't happen
+				log.Errorf("wrong type %T: %v", obj, obj)
+				return
+			}
+			c.mutex.RLock()
+			defer c.mutex.RUnlock()
+			for _, handler := range c.callbacks {
+				handler(crd.Name)
+			}
+		},
+		UpdateFunc: nil,
+		DeleteFunc: nil,
+	})
+
+	c.informer = crdMetadataInformer
+	return c
+}
+
+func (c *Controller) Run(stop <-chan struct{}) {
+	go c.informer.Run(stop)
```

Comment:
I am not sure this is appropriate. We use the shared informer factory, so I would expect us to use `client.RunAndWait`. 

We also are implicitly relying on crdclient to call this in kube/controller. IMO it would be more appropriate to:

1. Build crdWatcher in higher level initialization code.
2. Pass same crdWatcher to both components
3. In each component Run(), Add CRD handler. Then wait for crdWatcher.HasSynced(). Then call crdWatcher.List() and process all existing items

Passing same CRD watcher is not a hard requirement, they will use the same informer cache anyways so its not a high overhead. But (3) is important. Currently there is a race. We wait for the crdwatcher to sync, then we derive action. But AddCallbacks has two races:

1. It MUST be added before Run()... but I think because it uses a shared informer factory, even though we always add the callback before Run() is called, the informer may actually already be running. So when we call Run(), nothing happens. Then when we AddCallBack, it never gets events sent, since the event already happened.
2. Events are async, so HasSynced() can return true before the event are processed. This means in controller HasSynced(), we may see crdWatcher.HasSynced() but the MCS parts do not have CRD process yet

Reply:
>I am not sure this is appropriate. We use the shared informer factory, so I would expect us to use client.RunAndWait.

Yeah, the `Run` here can be removed. Originally I explicitly leave it here is to allow a separate start.

Reply:
For 1. we have made client.RunAndWait always been run at last, so it must be after `AddCallBack`.

>Events are async, so HasSynced() can return true before the event are processed. This means in controller HasSynced(), we may see crdWatcher.HasSynced() but the MCS parts do not have CRD process yet


IIUC this s not right. `HasSynced` returns true when all the items CB handlers has been run.

You can see the underlying implemet

```
// HasSynced returns true if an Add/Update/Delete/AddIfNotPresent are called first,
// or the first batch of items inserted by Replace() has been popped.
func (f *DeltaFIFO) HasSynced() bool {
	f.lock.Lock()
	defer f.lock.Unlock()
	return f.populated && f.initialPopulationCount == 0
}
```

Reply:
> IIUC this s not right. HasSynced returns true when all the items CB handlers has been run.

This is not accurate unfortunately... see https://github.com/kubernetes/kubernetes/issues/113763. They added a new thing in 1.27 to check this, but HasSynced does not.

Reply:
Thanks, i have got it

Reply:
Should we wait until 1.27 or continue doing a workaround